File: apps/cli/internal/defaults/defaults.go
package defaults

import (
	"os"
	"path/filepath"
	"strings"
)

type generatedDefaultsSpec struct {
	State generatedStateDefaults
}

type generatedStateDefaults struct {
	ProductionDirName    string
	DevelopmentDirName   string
	RunDirName           string
	ListenerInfoFileName string
}

type ResolvedDefaults struct {
	Runtime RuntimeDefaults
	State   StateDefaults
}

type RuntimeDefaults struct {
	Env string
}

type StateDefaults struct {
	RootDir                 string
	RunDir                  string
	NextopdListenerInfoPath string
}

func ResolveDefaultsFromEnv() ResolvedDefaults {
	env := resolveNextopEnv()
	stateRootDir := resolveStateRootDir(env)
	runDir := resolveRunDir(stateRootDir)

	return ResolvedDefaults{
		Runtime: RuntimeDefaults{
			Env: env,
		},
		State: StateDefaults{
			RootDir:                 stateRootDir,
			RunDir:                  runDir,
			NextopdListenerInfoPath: resolveListenerInfoPath(runDir),
		},
	}
}

func resolveNextopEnv() string {
	value := strings.ToLower(strings.TrimSpace(os.Getenv("NEXTOP_ENV")))
	switch value {
	case "dev", "development", "local":
		return "development"
	default:
		return "production"
	}
}

func resolveStateRootDir(env string) string {
	override := strings.TrimSpace(os.Getenv("NEXTOP_STATE_DIR"))
	if override != "" {
		return override
	}

	homeDir, err := os.UserHomeDir()
	if err != nil || strings.TrimSpace(homeDir) == "" {
		if env == "development" {
			return generatedDefaults.State.DevelopmentDirName
		}
		return generatedDefaults.State.ProductionDirName
	}

	dirName := generatedDefaults.State.ProductionDirName
	if env == "development" {
		dirName = generatedDefaults.State.DevelopmentDirName
	}

	return filepath.Join(homeDir, dirName)
}

func resolveRunDir(stateRootDir string) string {
	override := strings.TrimSpace(os.Getenv("NEXTOPD_RUN_DIR"))
	if override != "" {
		return override
	}

	return filepath.Join(stateRootDir, generatedDefaults.State.RunDirName)
}

func resolveListenerInfoPath(runDir string) string {
	override := strings.TrimSpace(os.Getenv("NEXTOPD_LISTENER_INFO_PATH"))
	if override != "" {
		return override
	}

	return filepath.Join(runDir, generatedDefaults.State.ListenerInfoFileName)
}

File: apps/desktop/src/main/defaults.ts
import { join } from "node:path";
import { homedir } from "node:os";
import { generatedDefaults } from "./generated/defaults.ts";

export interface DesktopResolvedDefaults {
  runtime: {
    env: "development" | "production";
  };
  state: {
    rootDir: string;
    logsDir: string;
    runDir: string;
    nextopdDBPath: string;
    nextopdListenerInfoPath: string;
    nextopdLogPath: string;
    desktopLogPath: string;
    nextopdPIDPath: string;
  };
  transport: {
    tcpAddr: string;
  };
  logging: {
    defaultLevel: "debug" | "info" | "warn" | "error";
    defaultOutput: "file" | "stdout" | "tee";
    maxSizeMB: number;
    maxBackups: number;
    maxAgeDays: number;
    maxTotalMB: number;
  };
}

export function initializeDesktopEnvironment(options?: {
  appVersion?: string;
  isPackaged?: boolean;
}): void {
  if (!process.env.NEXTOP_ENV?.trim()) {
    process.env.NEXTOP_ENV = options?.isPackaged ? "production" : "development";
  }

  const appVersion = options?.appVersion?.trim();
  if (!process.env.NEXTOP_APP_VERSION?.trim() && appVersion) {
    process.env.NEXTOP_APP_VERSION = appVersion;
  }
}

export function resolveDesktopDefaultsFromEnv(): DesktopResolvedDefaults {
  const env = resolveNextopEnv();
  const stateRootDir = resolveStateRootDir(env);
  const logsDir = resolveLogsDir(stateRootDir);
  const runDir = resolveRunDir(stateRootDir);

  return {
    runtime: {
      env
    },
    state: {
      rootDir: stateRootDir,
      logsDir,
      runDir,
      nextopdDBPath: resolveDBPath(stateRootDir),
      nextopdListenerInfoPath: resolveListenerInfoPath(runDir),
      nextopdLogPath: resolveDaemonLogPath(logsDir),
      desktopLogPath: resolveDesktopLogPath(logsDir),
      nextopdPIDPath: resolvePIDPath(runDir)
    },
    transport: {
      tcpAddr: resolveTCPAddr()
    },
    logging: {
      defaultLevel: generatedDefaults.logging.defaultLevel,
      defaultOutput: generatedDefaults.logging.defaultOutput,
      maxSizeMB: generatedDefaults.logging.maxSizeMB,
      maxBackups: generatedDefaults.logging.maxBackups,
      maxAgeDays: generatedDefaults.logging.maxAgeDays,
      maxTotalMB: generatedDefaults.logging.maxTotalMB
    }
  };
}

export function resolveNextopEnv(): "development" | "production" {
  return process.env.NEXTOP_ENV?.trim().match(/^(dev|development|local)$/i)
    ? "development"
    : "production";
}

function resolveStateRootDir(env: "development" | "production"): string {
  const override = process.env.NEXTOP_STATE_DIR?.trim();
  if (override) {
    return override;
  }

  const homeDir = homedir();
  const dirName =
    env === "development"
      ? generatedDefaults.state.developmentDirName
      : generatedDefaults.state.productionDirName;

  return join(homeDir, dirName);
}

function resolveLogsDir(stateRootDir: string): string {
  const override = process.env.NEXTOP_LOG_DIR?.trim();
  if (override) {
    return override;
  }

  return join(stateRootDir, generatedDefaults.state.logsDirName);
}
function resolveRunDir(stateRootDir: string): string {
  const override = process.env.NEXTOPD_RUN_DIR?.trim();
  if (override) {
    return override;
  }

  return join(stateRootDir, generatedDefaults.state.runDirName);
}

function resolveDBPath(stateRootDir: string): string {
  const override = process.env.NEXTOPD_DB_PATH?.trim();
  if (override) {
    return override;
  }

  return join(stateRootDir, generatedDefaults.state.dbFileName);
}

function resolveDaemonLogPath(logsDir: string): string {
  const override = process.env.NEXTOPD_LOG_PATH?.trim();
  if (override) {
    return override;
  }

  return join(logsDir, generatedDefaults.state.daemonLogFileName);
}

function resolveDesktopLogPath(logsDir: string): string {
  const override = process.env.NEXTOP_DESKTOP_LOG_PATH?.trim();
  if (override) {
    return override;
  }

  return join(logsDir, generatedDefaults.state.desktopLogFileName);
}

function resolvePIDPath(runDir: string): string {
  const override = process.env.NEXTOPD_PID_PATH?.trim();
  if (override) {
    return override;
  }

  return join(runDir, generatedDefaults.state.pidFileName);
}

function resolveListenerInfoPath(runDir: string): string {
  const override = process.env.NEXTOPD_LISTENER_INFO_PATH?.trim();
  if (override) {
    return override;
  }

  return join(runDir, generatedDefaults.state.listenerInfoFileName);
}

function resolveTCPAddr(): string {
  return (
    process.env.NEXTOPD_ADDR?.trim() ||
    generatedDefaults.transport.defaultTCPAddr
  );
}
